File: 2022/0122_20.有效的括号.c
/*
 * @lc app=leetcode.cn id=20 lang=c
 *
 * [20] 有效的括号
 */

// @lc code=start

/*
 * 今天2022.1.22是学习栈的2nd天
 * Python算法设计分析书里讲了栈的两个Python实现
 * 第一个是()匹配，第二个是()、[]、{}匹配
 * 拿这道题复习一下，依旧C、Python、C++都写一遍练语法熟悉度
 */

// 本题类同[1614]括号的最大嵌套深度

/*
栈的思维：
遇到左符号就push，遇到右符号就与栈顶元素top相匹配，不匹配返回False，否则就pop出top元素，继续循环
直到栈为空，就返回True
*/
bool isValid(char *s)
{
    int len = strlen(s);
    char stack[len + 1];
    // 或char *stack = (char*)malloc(len + 1);
    int top = 0; // top后期作为stack里的指针，所以不要为-1，否则数组越下界（-1<0)
    int flag = 1;
    // 题目说数组长度大于等于1，所以不用考虑空数组
    // 如果第一个符号就是右符号，直接返回false
    if (s[0] == ')' || s[0] == ']' || s[0] == '}' || len % 2 != 0)
    {
        return false;
    }

    for (int i = 0; i < len; ++i)
    {
        if (s[i] == '(' || s[i] == '[' || s[i] == '{')
        {
            stack[top] = s[i]; // 左符号入栈
            top++;
        }
        else
        {
            if (top > 0)
            {
                flag = matches(stack[top - 1], s[i]); // 前面top++了，所以要top - 1;另外要防止top-1越0下界
                if (flag == 0)
                    return false; // 符号不匹配，直接返回false
                else              // 符号匹配，top元素往前推一位
                {
                    top--; // 防止数组越下界出现Runtime Error
                }
            }
            else
                return false;
        }
    }
    if (top == 0)
        return true;
    else
        return false;
}

int matches(opens, closes)
{
    if (opens == '(' && closes == ')')
        return 1;
    else if (opens == '[' && closes == ']')
        return 1;
    else if (opens == '{' && closes == '}')
        return 1;
    else
        return 0;
}
// @lc code=end

File: 2022/0122_20.有效的括号.py
# @before-stub-for-debug-begin
from python3problem20 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
# 打算用Python算法设计书里的代码试一下
# LeetCode没有：
# from Pythonds.basic import Stack
# 建立class：Stack,并定义基本功能
# C中按题目输出返回是小写false，但是Python中false会报错，必须写成False

class Solution:
    def isValid(self, s: str) -> bool:
        ar = Stack()

        balanced = True
        index = 0

        while index < len(s) and balanced:
            symbol = s[index]
            if symbol in "([{":
                ar.push(symbol) 
            else:
                if ar.isEmpty():   # 要防止出现空列表后还pop，空列表直接返回flase
                    balanced = False 
                else:
                    top = ar.pop()   # 列表非空就把左符号pop出，pop返回刚pop出的元素，并赋值给top
                    if matches(top, symbol) == 0:  # 将top元素与symbol元素匹配
                        balanced = False   # 如果不匹配balanced就为false，否则保持true
            
            index = index + 1
        
        # return ar.isEmpty() and balanced也可以
        if balanced and ar.isEmpty():  # isEmpty():假如A为空的话，返回的值是1 # 两者取交集
            return True
        else:   # if后一定不要忘记写else
            return False

def matches(open, close):  # 写这个匹配函数明显Python比C方便
    opens = "([{"
    closers = ")]}"

    return opens.index(open) == closers.index(close)

class Stack(): 
    def __init__(self):
        self.items = []  #这一段不可少，否则Runtime Error

    def push(self, item):
        self.items.append(item)
        
    def pop(self):
        return self.items.pop()  # pop返回刚才pop出的元素
    
    def isEmpty(self):
        return self.items == []  # 空返回1，反之返回0 

# @lc code=end

